### PR TITLE
Split CI YAML into smaller templates

### PR DESCRIPTION
**Pull request by zivkan**
## Bug

Fixes: https://github.com/NuGet/Client.Engineering/issues/757

Regression? no

## Description

In the past, our CI YAML was one, very long file. I found it hard to work with. A few weeks/months ago, two of our jobs were moved to templates. I think it's better, so in this PR I did some refactoring:
* Every "job" now has its own template.
* Moved the list of jobs to `eng/pipelines`, because it's more in line with dnceng's Arcade, and I want us to slowly move towards being more in-line with the rest of the dotnet org.
* set up `build/vsts_build.yaml` to be nothing more than an import of our `pipeline.yml`.
* Created `eng/pipelines/official.yml` and `eng/pipelines/pull_requests.yml`, which are also nothing more than imports of `pipeline.yml`, because [runtime parameters](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/runtime-parameters?view=azure-devops&tabs=script) cannot use build variables, so if we want to start using this in the future, we need a different yaml entry point for each pipeline.
   * These will not be used for now, but it ensures that if/when in the future we want to start using them, anyone with open pull requests should hopefully already be using a commit where the new yaml paths already exist. We will, however, need to cherry-pick this PR onto old release branches in order to service them.

## PR Checklist

- [X] PR has a meaningful title
- [X] PR has a linked issue.
- [X] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [X] N/A (Infrastructure)

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [X] N/A

### CODE REVIEW

**Comment by zivkan** [comment]
@kartheekp-ms I know you're working on something that will need a yaml file change, and this PR is going to cause merge conflicts.

**Review by donnie-msft — APPROVED** [review]
Much easier to digest our script now! 

> set up build/vsts_build.yaml to be nothing more than an import of our pipeline.yml.

I'm assuming our CI still shows the specific jobs/task names when failures occur.

File: eng/pipelines/templates/Tests_On_Linux.yml [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    disableAutoCwd: "true"
+    cwd: "$(Build.Repository.LocalPath)"
+  condition: "and(succeeded(), not(eq(variables['IsOfficialBuild'], 'true')))"
+
+- task: PublishTestResults@2
+  displayName: "Publish Test Results"
+  inputs:
+    testRunner: "XUnit"
+    testResultsFiles: "*.xml"
+    testRunTitle: "NuGet.Client Tests On Linux"
```
Comment (kartheekp-ms) [inline eng/pipelines/templates/Tests_On_Linux.yml:50]: Comparing `eng/pipelines/templates/Tests_On_Linux.yml` with `eng/pipelines/templates/Tests_On_Mac.yml` which are almost identical except `testRunTitle`which  will vary based on the OS.  Can we somehow use same `.yml` for `Linux` and `Mac` Tests? The same comment applies to [`Line no:70`](https://github.com/NuGet/NuGet.Client/pull/3883/files#diff-9c049d86680c1cf227b96f4899b00768747c5bff89b90f3fc3f6b530fea8bb48R70) in this `.yml` file.

**Comment by zivkan** [comment]
> I'm assuming our CI still shows the specific jobs/task names when failures occur

Yes, you can look at the CI build for this PR and see all the jobs.

Much like MSBuild, or C language macros, Azure Pipelines "pre-processes" the file (create a single representation with all imports inlined), so the runtime doesn't know any different. It's just for our own organisation.

File: eng/pipelines/templates/Tests_On_Linux.yml (hunk 1 above)
Comment (kartheekp-ms) [reply eng/pipelines/templates/Tests_On_Linux.yml:50]: `Tests_On_Mac.yml` file has additional `"Download NuGet.CommandLine.Test artifacts"` task compare to `Tests_On_Linux.yml`

Comment (zivkan) [reply eng/pipelines/templates/Tests_On_Linux.yml:50]: yes we can, but it's outside of the scope of this PR, which is just to split them into more managable pieces. In fact, this split is what makes the comparison of the Mac/Linux jobs so easy. I'd like to get this split in ASAP, so we can start making other improvements with less risk of merge conflicts.

**Review by kartheekp-ms — APPROVED** [review]

**Review by erdembayar — APPROVED** [review]

**Comment by donnie-msft** [comment]
> > I'm assuming our CI still shows the specific jobs/task names when failures occur
> 
> Yes, you can look at the CI build for this PR and see all the jobs.
> 
> Much like MSBuild, or C language macros, Azure Pipelines "pre-processes" the file (create a single representation with all imports inlined), so the runtime doesn't know any different. It's just for our own organisation.

Ok, but this says it's showing results "from a previous attempt". Is that expected?
https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4434310&view=logs&j=a1762d43-9ec6-55e8-af8b-c0d9842bd83b
image

**Comment by zivkan** [comment]
That's normal when there was a flaky test in a different job (Windows func tests, IsDesktop), and using "retry failed jobs". If you expand "Attempt 1", you should see more clearly.

Perhaps AzDO changed their UI recently, but if so it's coincidental and unrelated to this PR.

**Pull request merged**